src/python/pants/backend/python/rules/inject_init.py

```diff
@@ -26,33 +34,47 @@ class InitInjectedSnapshot:
 async def inject_missing_init_files(request: InjectInitRequest) -> InitInjectedSnapshot:
     """Ensure that every package has an `__init__.py` file in it.
 
-    This will preserve any `__init__.py` files already in the input snapshot.
+    This will first use any `__init__.py` files in the input snapshot, then read from the filesystem
+    to see if any exist but are not in the snapshot, and finally will create empty files.
     """
-    snapshot = request.sources_snapshot
-    missing_init_files = identify_missing_init_files(snapshot.files)
-    if not missing_init_files:
-        return InitInjectedSnapshot(snapshot)
+    original_missing_init_files = identify_missing_init_files(request.snapshot.files)
+    if not original_missing_init_files:
+        return InitInjectedSnapshot(request.snapshot)
 
+    missing_init_files = original_missing_init_files
     if not request.sources_stripped:
         # Get rid of any identified-as-missing __init__.py files that are not under a source root.
-        missing_init_files_list = list(missing_init_files)
         optional_src_roots = await MultiGet(
             Get(OptionalSourceRoot, SourceRootRequest, SourceRootRequest.for_file(init_file))
-            for init_file in missing_init_files_list
+            for init_file in original_missing_init_files
         )
-        for optional_src_root, init_file in zip(optional_src_roots, missing_init_files_list):
-            # If the identified-as-missing __init__.py file is above or at a source root, remove it.
-            if (
-                optional_src_root.source_root is None
-                or optional_src_root.source_root.path == os.path.dirname(init_file)
-            ):
-                missing_init_files.remove(init_file)
 
+        def is_under_source_root(init_file: str, optional_src_root: OptionalSourceRoot) -> bool:
+            return (
+                optional_src_root.source_root is not None
+                and optional_src_root.source_root.path != os.path.dirname(init_file)
+            )
+
+        missing_init_files = FrozenOrderedSet(
+            init_file
+            for init_file, optional_src_root in zip(original_missing_init_files, optional_src_roots)
+            if is_under_source_root(init_file, optional_src_root)
+        )
+
+    discovered_inits_snapshot = await Get(Snapshot, PathGlobs(missing_init_files))
     generated_inits_digest = await Get(
         Digest,
-        InputFilesContent(FileContent(path=fp, content=b"") for fp in sorted(missing_init_files)),
+        InputFilesContent(
+            FileContent(fp, b"# Generated `__init__.py` file.")
```

Comment:
```suggestion
            FileContent(fp, b"# Generated by Pants to ensure that this package is importable.")
```

Reply:
We already know this is an `__init__.py` file so stating so in the content is redundant.